Sema: Allow opening existential with a warning for rdar://141962317

AnthonyLatsis opened the pull request:
We now correctly refuse to open the existential in the following case due to the invariant `T` in the second parameter, which broke some code internally. Opening the existential in this precise case happens to be safe, so this patch adds a temporary exception to the rules and pairs it with a warning so folks can spot and fix the problem.

```swift
struct S<each T> {}
protocol P {}
func open<T: P>(_: T.Type, _: S<T>? = nil) {}
let meta: any P.Type
open(meta)
```

AnthonyLatsis committed:
Sema: Allow opening existential with a warning for rdar://141962317

We now correctly refuse to open the existential in the following case
due to the invariant `T` in the second parameter, which broke some code
internally. Opening the existential in this precise case happens to be
safe, so this patch adds a temporary exception to the rules and pairs it
with a warning so folks can spot and fix the problem.

```swift
struct S<each T> {}
protocol P {}
func open<T: P>(_: T.Type, _: S<T>? = nil) {}
let meta: any P.Type
open(meta)
```

AnthonyLatsis committed:
[NFC] Diag: Rename `error_in_future_swift_version` wrappers

Be clear that these are about language mode.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -9576,3 +9576,12 @@ bool IncorrectSlabLiteralCount::diagnoseAsError() {
+
+bool WarnAboutExistentialOpenedForCallArgumentUntilFutureRelease::
+    diagnoseAsError() {
+  auto overload = *getCalleeOverloadChoiceIfAvailable(getLocator());
```
Nit: we have `getOverloadChoice` for this which force unwraps inside.

AnthonyLatsis replied:
This also looks through to the callee locator, though.

xedin replied:
Oh, I missed that. Let's add a variant without `IfAvailable` to solution just like `getOverloadChoice`.

AnthonyLatsis replied:
Without a forwarding method on `FailureDiagnostic`, right?

AnthonyLatsis committed:
DiagnosticEngine: Introduce `warnUntilFutureSwiftCompiler()`

This establishes an resusable approach to staging fixes for potentially
harmful miscompilations that we do not want to keep behind a language
mode.
